File: app/src/main/java/com/example/parliamentvoiceapp/viewmodel/VoiceViewModel.java
package com.example.parliamentvoiceapp.viewmodel;

import android.app.Application;
import android.util.Log;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import com.example.parliamentvoiceapp.audio.SpeechManager;
import com.example.parliamentvoiceapp.network.ApiService;
import com.example.parliamentvoiceapp.network.AskRequest;
import com.example.parliamentvoiceapp.network.AskResponse;
import com.example.parliamentvoiceapp.network.RetrofitClient;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class VoiceViewModel extends AndroidViewModel {

    private static final String TAG = "VoiceViewModel";
    private SpeechManager speechManager;
    private MutableLiveData<String> correctedText = new MutableLiveData<>();

    // Simple dictionary for autocorrect (expand as needed)
    private static final Map<String, String> DICTIONARY = new HashMap<>();

    static {
        DICTIONARY.put("teh", "the");
        DICTIONARY.put("recieve", "receive");
        DICTIONARY.put("adress", "address");
        // add more common misspellings
    }

    public VoiceViewModel(@NonNull Application application) {
        super(application);
        speechManager = new SpeechManager(application.getApplicationContext());

        // Observe recognized text and autocorrect it
        speechManager.getRecognizedText().observeForever(text -> {
            if (text != null) {
                if (text.isEmpty()) {
                    correctedText.postValue("");
                } else {
                    String autoCorrected = autocorrect(text);
                    correctedText.postValue(autoCorrected);
                }
            }
        });
    }

    /**
     * Basic autocorrect method that replaces common misspelled words from the dictionary.
     * Can be improved with more sophisticated algorithms or Android spellchecker.
     */
    private String autocorrect(String input) {
        String[] words = input.split("\\s+");
        StringBuilder corrected = new StringBuilder();

        for (String word : words) {
            String lowerWord = word.toLowerCase();
            if (DICTIONARY.containsKey(lowerWord)) {
                corrected.append(DICTIONARY.get(lowerWord));
            } else {
                corrected.append(word);
            }
            corrected.append(" ");
        }

        return corrected.toString().trim();
    }

    public static class ChatMessage {
        public String text;
        public boolean isUser;
        public ChatMessage(String text, boolean isUser) {
            this.text = text;
            this.isUser = isUser;
        }
    }

    private MutableLiveData<List<ChatMessage>> chatHistory = new MutableLiveData<>(new ArrayList<>());

    public LiveData<List<ChatMessage>> getChatHistory() {
        return chatHistory;
    }

    public void submitQuery(String query) {
        if (query == null || query.trim().isEmpty()) return;
        
        List<ChatMessage> currentList = new ArrayList<>(chatHistory.getValue());
        currentList.add(new ChatMessage(query, true));
        chatHistory.postValue(currentList);
        
        // Clear the current input text since it's already sent
        correctedText.postValue("");
        
        // Call the backend API
        ApiService apiService = RetrofitClient.getApiService();
        AskRequest request = new AskRequest(query);
        
        apiService.askQuestion(request).enqueue(new Callback<AskResponse>() {
            @Override
            public void onResponse(Call<AskResponse> call, Response<AskResponse> response) {
                if (response.isSuccessful() && response.body() != null) {
                    String answer = response.body().getAnswer();
                    addBotResponse(answer);
                } else {
                    Log.e(TAG, "API Error: " + response.code());
                    addBotResponse("Error: Unable to get response from server. (Code: " + response.code() + ")");
                }
            }

            @Override
            public void onFailure(Call<AskResponse> call, Throwable t) {
                Log.e(TAG, "API Failure: " + t.getMessage());
                addBotResponse("Error: Network failure. Please check your connection.");
            }
        });
    }

    private void addBotResponse(String responseText) {
        List<ChatMessage> updatedList = new ArrayList<>(chatHistory.getValue());
        updatedList.add(new ChatMessage(responseText, false));
        chatHistory.postValue(updatedList);
    }

    public void clearChat() {
        chatHistory.postValue(new ArrayList<>());
    }

    public LiveData<String> getCorrectedText() {
        return correctedText;
    }

    public void updateText(String newText) {
        correctedText.postValue(newText);
    }

    public LiveData<Boolean> getIsListening() {
        return speechManager.getIsListening();
    }

    public LiveData<Float> getRmsDb() {
        return speechManager.getRmsDb();
    }

    public void startListening() {
        speechManager.startListening();
    }

    public void stopListening() {
        speechManager.stopListening();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        speechManager.destroy();
    }
}

File: app/src/main/kotlin/com/example/parliamentvoiceapp/viewmodel/VoiceViewModel.kt
package com.example.parliamentvoiceapp.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.example.parliamentvoiceapp.audio.SpeechManager
import com.example.parliamentvoiceapp.audio.TTSManager
import com.example.parliamentvoiceapp.db.AppDatabase
import com.example.parliamentvoiceapp.db.ChatSession
import com.example.parliamentvoiceapp.db.LocalChatMessage
import com.example.parliamentvoiceapp.network.AskRequest
import com.example.parliamentvoiceapp.network.HistoryItem
import com.example.parliamentvoiceapp.network.RetrofitClient
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

class VoiceViewModel(application: Application) : AndroidViewModel(application) {

    private val db = AppDatabase.getDatabase(application)
    private val chatDao = db.chatDao()
    private val speechManager = SpeechManager(application.applicationContext)
    private val ttsManager = TTSManager(application.applicationContext)
    
    private val _correctedText = MutableLiveData<String>("")
    val correctedText: LiveData<String> = _correctedText

    private val _chatHistory = MutableLiveData<List<ChatMessage>>(emptyList())
    val chatHistory: LiveData<List<ChatMessage>> = _chatHistory

    private var currentSessionId: Long? = null
    private var sessionObservationJob: Job? = null

    private val _allSessions = MutableLiveData<List<ChatSession>>(emptyList())
    val allSessions: LiveData<List<ChatSession>> = _allSessions

    data class ChatMessage(
        val text: String,
        val isUser: Boolean,
        val isError: Boolean = false
    )

    init {
        speechManager.recognizedText.observeForever { text ->
            if (text != null) _correctedText.postValue(text)
        }

        viewModelScope.launch {
            chatDao.getAllSessions().collectLatest {
                _allSessions.postValue(it)
            }
        }
    }

    fun loadSession(sessionId: Long) {
        // Cancel previous session observation to prevent data mixing
        sessionObservationJob?.cancel()
        currentSessionId = sessionId
        
        sessionObservationJob = viewModelScope.launch {
            chatDao.getMessagesForSession(sessionId).collectLatest { messages ->
                _chatHistory.postValue(messages.map { 
                    ChatMessage(it.text, it.isUser, it.isError) 
                })
            }
        }
    }

    fun startNewSession() {
        sessionObservationJob?.cancel()
        currentSessionId = null
        _chatHistory.value = emptyList()
        _correctedText.value = ""
    }

    fun submitQuery(query: String) {
        if (query.isBlank()) return

        viewModelScope.launch {
            // 1. Ensure Session exists
            if (currentSessionId == null) {
                val newId = chatDao.insertSession(ChatSession(title = query))
                currentSessionId = newId
                // Start observing the new session immediately
                loadSession(newId)
            }

            val sessionId = currentSessionId!!

            // 2. Immediate UI Update for User Message
            val updatedListWithUser = (_chatHistory.value ?: emptyList()).toMutableList()
            updatedListWithUser.add(ChatMessage(query, true))
            _chatHistory.value = updatedListWithUser
            _correctedText.value = ""
            ttsManager.stop()

            // 3. Background Save
            launch { chatDao.insertMessage(LocalChatMessage(sessionId = sessionId, text = query, isUser = true)) }

            // 4. Prepare Context for Backend
            val historyItems = updatedListWithUser
                .filter { !it.isError }
                .map { HistoryItem(role = if (it.isUser) "user" else "assistant", content = it.text) }
                .dropLast(1)

            // 5. API Call
            try {
                val response = RetrofitClient.apiService.askQuestion(
                    AskRequest(query = query, history = historyItems)
                )
                val answer = response.answer ?: "क्षमा करें, कोई उत्तर नहीं मिला।"
                
                // 6. Immediate UI Update for Assistant Bubble
                val updatedListWithAI = _chatHistory.value?.toMutableList() ?: mutableListOf()
                updatedListWithAI.add(ChatMessage(answer, false))
                _chatHistory.postValue(updatedListWithAI)

                // 7. Speak and Save in Parallel
                ttsManager.speak(answer)
                launch { chatDao.insertMessage(LocalChatMessage(sessionId = sessionId, text = answer, isUser = false)) }
                
            } catch (e: Exception) {
                val errorMsg = "Network error. Please try again."
                val updatedListWithError = _chatHistory.value?.toMutableList() ?: mutableListOf()
                updatedListWithError.add(ChatMessage(errorMsg, false, true))
                _chatHistory.postValue(updatedListWithError)
                ttsManager.speak("नेटवर्क त्रुटि।")
            }
        }
    }

    fun updateText(newText: String) = _correctedText.postValue(newText)
    fun clearChat() = startNewSession()

    val isListening: LiveData<Boolean> = speechManager.isListening
    val rmsDb: LiveData<Float> = speechManager.rmsDb

    fun startListening() {
        ttsManager.stop()
        speechManager.startListening()
    }
    
    fun stopListening() = speechManager.stopListening()

    override fun onCleared() {
        super.onCleared()
        speechManager.destroy()
        ttsManager.shutdown()
    }
}